app/dao/notification_history_dao.py

```diff
@@ -6,13 +6,27 @@
 def delete_notification_history_older_than_datetime(older_than_datetime, query_limit=50000):
     current_app.logger.info("Beginning to delete notification_history older than %s", older_than_datetime)
 
+    # We want to run this function concurrently, ie having several workers running this same task at the same time
+    # To ensure workers don't try and delete the same notifications (which will cause them to stop and exit)
+    # we get them to pick a random sample of notifications
+    # In production, to get 50,000 results we only need a random sample of about 0.05% of the table to give us
+    # 50,000 rows
+    # However, in other environments, including test, we don't have many notifications in the notification_history
+    # table and therefore need to get all of them (and don't need to worry about performance)
+    if current_app.config["NOTIFY_ENVIRONMENT"] == "production":
+        tablesample_percentage = 0.05
```

Comment:
will this value not change as notification_history shrinks?

Reply:
Good point Pea. I hadn't given this enough thought, or made it clear my thinking in the commit message and I should have.

After thinking it through, this value is OKish and would do the job but could maybe be better.

I checked and in our prod DB we have just over 1 billion notifications older than 2023-01-01 (ie these are the ones to not be deleted).

If we took 0.05 percent of 1 billion, it is 500,000. This means that tablesample will find *at least* 500,000 notifications. It will find more to start with whilst there are closer to 5 billion notifications, and at the end will find about 500,000 as we have removed the number of notifications. Crucially, we still have the `limit` clause on the end which will then pick only 50,000 of the ones picked by `tablesample`.

So 0.05 looks sufficiently large enough.

However, if we picked 0.005%, it might be more efficient. 0.005% of 1 billion is 50,000. So we would expect at least 50,000 notifications to be found each time still.

So the current value of 0.05% currently in the commit wouldn't encounter any problems, but we could potentially change it to 0.005% which I assume would result in quicker sql.